apps/aecore/lib/aecore/peers/worker.ex

```diff
@@ -162,28 +168,42 @@ defmodule Aecore.Peers.Worker do
   ## Internal functions
   defp add_peer(uri, state) do
     %{peers: peers} = state
-    if Map.has_key?(peers, uri) do
+    state_has_uri =
+      peers
+        |> Map.values()
+        |> Enum.map(fn(%{uri: uri}) -> uri end)
+        |> Enum.member?(uri)
+
+    if state_has_uri do
       Logger.debug(fn ->
         "Skipped adding #{uri}, already known" end)
       {:reply, {:error, "Peer already known"}, state}
     else
       case check_peer(uri, get_peer_nonce()) do
         {:ok, info} ->
-          if should_a_peer_be_added(map_size(peers)) do
-            peers_update1 =
-            if map_size(peers) >= @peers_max_count do
-                random_peer = Enum.random(Map.keys(peers))
-                Logger.debug(fn -> "Max peers reached. #{random_peer} removed" end)
-                Map.delete(peers, random_peer)
-              else
-                peers
-              end
-            updated_peers = Map.put(peers_update1, uri, info.current_block_hash)
-            Logger.info(fn -> "Added #{uri} to the peer list" end)
-            {:reply, :ok, %{state | peers: updated_peers}}
+          if(!Map.has_key?(peers, info.peer_nonce)) do
```

Comment:
This check should be inside check_peer function. The purpose of `check_peer` function is to separate all of our checks from the main logic. (if inside if, inside if, inside... doesn't "look nice")

Reply:
We need the state for this check and we don't have it in check_peer, passing it just for that seems unnecessary

Reply:
maybe we just create a separate function for it

Reply:
It's definitely better now. We could pass `peers` to `check_peer` and move the check, but I think it's good enough now.